runtime/interpreter/interpreter.go

```diff
@@ -1905,10 +1988,88 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
 			return ConvertAddress(interpreter, value, locationRange)
 		}
 
+	case *sema.ConstantSizedType, *sema.VariableSizedType:
+		if arrayValue, isArray := value.(*ArrayValue); isArray && !valueType.Equal(unwrappedTargetType) {
+
+			arrayStaticType := interpreter.convertStaticType(arrayValue.StaticType(interpreter), unwrappedTargetType).(ArrayStaticType)
+			targetElementType := interpreter.MustConvertStaticToSemaType(arrayStaticType.ElementType())
+
+			values := make([]Value, 0, arrayValue.Count())
+
+			arrayValue.Iterate(interpreter, func(v Value) bool {
+				valueType := interpreter.MustConvertStaticToSemaType(v.StaticType(interpreter))
+				values = append(values, interpreter.convert(v, valueType, targetElementType, locationRange))
+				return true
+			})
+
+			return NewArrayValue(
+				interpreter,
+				locationRange,
+				arrayStaticType,
+				arrayValue.GetOwner(),
+				values...,
+			)
+		}
+
+	case *sema.DictionaryType:
+		if dictValue, isDict := value.(*DictionaryValue); isDict && !valueType.Equal(unwrappedTargetType) {
+
+			dictStaticType := interpreter.convertStaticType(dictValue.StaticType(interpreter), unwrappedTargetType).(DictionaryStaticType)
+			targetKeyType := interpreter.MustConvertStaticToSemaType(dictStaticType.KeyType)
+			targetValueType := interpreter.MustConvertStaticToSemaType(dictStaticType.ValueType)
+
+			values := make([]Value, 0, dictValue.Count()*2)
+
+			dictValue.Iterate(interpreter, func(key, value Value) bool {
+				keyType := interpreter.MustConvertStaticToSemaType(key.StaticType(interpreter))
+				valueType := interpreter.MustConvertStaticToSemaType(value.StaticType(interpreter))
+
+				convertedKey := interpreter.convert(key, keyType, targetKeyType, locationRange)
+				convertedValue := interpreter.convert(value, valueType, targetValueType, locationRange)
+
+				values = append(values, convertedKey, convertedValue)
+				return true
+			})
+
+			return NewDictionaryValue(
+				interpreter,
+				locationRange,
+				dictStaticType,
+				values...,
+			)
+		}
+
+	case *sema.CapabilityType:
+		if !valueType.Equal(unwrappedTargetType) && unwrappedTargetType.BorrowType != nil {
+			targetBorrowType := unwrappedTargetType.BorrowType.(*sema.ReferenceType)
+
+			switch capability := value.(type) {
+			case *PathCapabilityValue:
+				valueBorrowType := capability.BorrowType.(ReferenceStaticType)
+				borrowType := interpreter.convertStaticType(valueBorrowType, targetBorrowType)
+				return NewPathCapabilityValue(
+					interpreter,
+					capability.Address,
+					capability.Path,
+					borrowType,
+				)
+			case *IDCapabilityValue:
+				valueBorrowType := capability.BorrowType.(ReferenceStaticType)
+				borrowType := interpreter.convertStaticType(valueBorrowType, targetBorrowType)
+				return NewIDCapabilityValue(
+					interpreter,
+					capability.ID,
+					capability.Address,
+					borrowType,
+				)
```

Comment:
Maybe panic with an unreachable error in the default case (unsupported capability value)